### PR TITLE
Masking improvements for MultiHeadAttention layer

### PR DESCRIPTION
**Pull request by maxeonyx**
The MultiHeadAttention (MHA) layer currently provides users with only one parameter `attention_mask` to use for masking (which is a N by M matrix mask where N and M are the lengths of the Key and Query tensors respectively).

I ran into an issue where this parameter is not sufficient, and it is necessary to additionally multiply the outputs of the MHA layer by a mask again, in order to prevent information leaking through. See the quoted text for my original summary.

After discussion with @fchollet, the correct course of action is to add 3 additional masking parameters to the MHA layer: `key_mask`, `query_mask` and `value_mask`. 

The most important change to fix my original issue is the addition of `value_mask`. When provided, `value_mask` should be multiplied against the `value` tensor before applying the attention operation. This will prevent very small floating point values from leaking through the MHA layer. A network can learn a setup to use these very small values, which manifests as inexplicably good performance. 

When provided, `key_mask` and `query_mask` should be used to compute `attention_mask` (e.g. with a broadcasted `tf.logical_and`). Providing either or both `key_mask` and `query_mask` should be mutually exclusive with providing `attention_mask`. Additionally, (not sure about this, or whether it is important) the implementation of the attention operation may be able to avoid materializing an N by M `attention_mask` when only `key_mask` or `query_mask` is provided.

Adding all three parameters means the MHA layer can be compaatible with Keras mask propagation. The values of `key_mask`, `query_mask` and `value_mask` can default to `x._keras_mask` for the respective tensor.

The original PR was only to change the documentation. Original text:

> Hi,
> 
> I discovered while working on an MNIST attention demo, that a network is able to learn adequate reconstruction performance on MNIST test set, even when the entire input is masked out.
> 
> When masking, before the softmax operation in the MultiHeadAttention layer the masked inputs are set to -1E9. This correctly results in masked value vectors having a negligible contribution to the final sum.
> 
> However, the network is able to learn to distinguish between the resulting *very small* floating point numbers of the masked out values, which appears as inexplicable good performance.
> 
> The fix (for self attention) is to multiply the attention outputs by the mask. This PR just puts a warning in the docs, but I encourage someone to think of a better solution.

### CODE REVIEW

**Comment by maxeonyx** [comment]
@googlebot I signed it!

**Comment by fchollet** [comment]
> However, the network is able to learn to distinguish between the resulting very small floating point numbers of the masked out values, which appears as inexplicable good performance.

This is not applicable unless the *entire* input is masked, no? Of course selective downscaling does not actually downscale in relative terms if the entire input gets uniformly downscaled. So it does not appear to be an issue in any real scenario.

**Comment by maxeonyx** [comment]
@fchollet In my project it occurs with any amount of masked out inputs - but the "everything masked out" case demonstrates that it is in fact this effect that is causing the too-good performance. If you would like, I can share the project, or put together a minimum example.

**Comment by fchollet** [comment]
Interesting ideas! Perhaps we should add 3 arguments, `key_mask`, `value_mask`, and `query_mask`, mutually exclusive with `attention_mask`. What would that look like?

**Comment by maxeonyx** [comment]
Yes - actually you're right it should be all 4 arguments. `attention_mask` can (typically) be computed from `key_mask` and `query_mask`. In most use cases, `value_mask` would usually be the same as `key_mask`, assuming we do the multiplication by the mask before the attention operation. If the mask is applied _after_ the attention operation, it would typically be the same as `query mask`. But, I think doing it _before_ is better. 

`attention_mask` would be mutually exclusive with both `key_mask` and `query_mask`, but not `value_mask`. In fact just the addition of `value_mask` would be a good fix to this issue.

**Comment by fchollet** [comment]
Thanks for the detailed breakdown. Sounds good -- then let's add the three arguments `key_mask`, `value_mask`, and `query_mask`, while keeping `attention_mask` for backwards compatibility. `key_mask` would default to `value_mask` if `key` was not specified. And all three arguments would default to `x._keras_mask` for their respective tensor.

**Comment by maxeonyx** [comment]
@fchollet Thanks! Updating this PR seems appropriate. To be clear, who will be performing the work? I see @gbaned has self-assigned it, does this mean they have volunteered? Alternatively, I could do it, but I haven't contributed to Keras before and the work might take me a while.

I can change the title and summary of the PR, but can the branch / commits be modified by someone else?

**Comment by fchollet** [comment]
> Alternatively, I could do it, but I haven't contributed to Keras before and the work might take me a while.

If you're willing to do it, please do. We'll provide you with any guidance or help you need.

**Comment by maxeonyx** [comment]
@fchollet Ok, I will give it a go

**Comment by fchollet** [comment]
@maxeonyx have you been able to look into it?

**Comment by maxeonyx** [comment]
@fchollet I haven't done any work on this - but it's still something I'd like to do if I have the time. Would you perhaps like me to close this and change it into an issue instead of a PR?

**Comment by fchollet** [comment]
> Would you perhaps like me to close this and change it into an issue instead of a PR?

Up to you -- we can keep the PR open. You can close it if you want. Good to know you still plan to work on the project!

**Comment by gbaned** [comment]
@maxeonyx Can you please resolve conflicts? Thanks!

**Comment by maxeonyx** [comment]
Hi - I will create an issue and/or a new PR if I find the time to do this work, but for now I think this is better closed.
